// @backstage/backend-defaults/dist/cjs/helpers-D2f1CG0o.cjs.js
'use strict';

function createCredentialsWithServicePrincipal(sub, token, accessRestrictions) {
  return {
    $$type: "@backstage/BackstageCredentials",
    version: "v1",
    token,
    principal: {
      type: "service",
      subject: sub,
      accessRestrictions
    }
  };
}
function createCredentialsWithUserPrincipal(sub, token, expiresAt) {
  return {
    $$type: "@backstage/BackstageCredentials",
    version: "v1",
    token,
    expiresAt,
    principal: {
      type: "user",
      userEntityRef: sub
    }
  };
}
function createCredentialsWithNonePrincipal() {
  return {
    $$type: "@backstage/BackstageCredentials",
    version: "v1",
    principal: {
      type: "none"
    }
  };
}
function toInternalBackstageCredentials(credentials) {
  if (credentials.$$type !== "@backstage/BackstageCredentials") {
    throw new Error("Invalid credential type");
  }
  const internalCredentials = credentials;
  if (internalCredentials.version !== "v1") {
    throw new Error(
      `Invalid credential version ${internalCredentials.version}`
    );
  }
  return internalCredentials;
}

exports.createCredentialsWithNonePrincipal = createCredentialsWithNonePrincipal;
exports.createCredentialsWithServicePrincipal = createCredentialsWithServicePrincipal;
exports.createCredentialsWithUserPrincipal = createCredentialsWithUserPrincipal;
exports.toInternalBackstageCredentials = toInternalBackstageCredentials;
