// @backstage/backend-defaults/dist/rootConfig.cjs.js
'use strict';

var createConfigSecretEnumerator = require('./cjs/createConfigSecretEnumerator-DShyoWWL.cjs.js');
var backendPluginApi = require('@backstage/backend-plugin-api');
var configLoader = require('@backstage/config-loader');
require('@manypkg/get-packages');

const rootConfigServiceFactoryWithOptions = (options) => backendPluginApi.createServiceFactory({
  service: backendPluginApi.coreServices.rootConfig,
  deps: {},
  async factory() {
    const source = configLoader.ConfigSources.default({
      argv: options?.argv,
      remote: options?.remote,
      watch: options?.watch
    });
    console.log(`Loading config from ${source}`);
    return await configLoader.ConfigSources.toConfig(source);
  }
})();
const rootConfigServiceFactory = Object.assign(
  rootConfigServiceFactoryWithOptions,
  rootConfigServiceFactoryWithOptions()
);

exports.createConfigSecretEnumerator = createConfigSecretEnumerator.createConfigSecretEnumerator;
exports.rootConfigServiceFactory = rootConfigServiceFactory;
